### PR TITLE
[IoT] Adds T2 Device Provisioning Service 2021-10-01 Dataplane Service SDK

### PR DESCRIPTION
**Pull request by c-ryan-k**
# Description

Adds Track 2 Service (Dataplane) SDK for Azure IoT Hub Device Provisioning Service
API spec is here: 
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json

API Review link: https://apiview.dev/Assemblies/Review/90240b0f26d946da9472062c5b6e72e3

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/iothub/azure-iot-deviceprovisioningservice/README.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
+# Azure IoT Hub Device Provisioning Service client library for Python
+
+The IoT Hub Device Provisioning Service (DPS) is a helper service for IoT Hub that enables zero-touch, just-in-time provisioning to the right IoT hub without requiring human intervention, allowing customers to provision millions of devices in a secure and scalable manner.
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/README.md:3]: tiny nit,  "...just-in-time provisioning to the right IoT **H**ub without require human ...."

File: sdk/iothub/azure-iot-deviceprovisioningservice/README.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        "registrationId": "<enrollment_id>"
+        "attestation": {
+            "type": "symmetricKey",
+        },
+    }
+)
+```
+
+### Create an enrollment group
+Create an x509 enrollment group to provision an individual device using x509 attestation.
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/README.md:121]: "Create an x509 enrollment group to provision **one or more** devices using x509 attestation." ?

File: sdk/iothub/azure-iot-deviceprovisioningservice/README.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+# Create x509 enrollment group
+client.enrollment_groups.create_or_update(
+    id="<enrollment_group_id>",
+    enrollment_group={
+        "enrollmentGroupId": "<enrollment_group_id>",
+        "attestation": {
+            "type": "x509",
+            "x509": {
+                "signingCertificates": {
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/README.md:140]: `signingCertificates` where its commonly used to store intermediate certificates, what about another example that makes use of `caReferences` such that the enrollment group will ensure whatever certificate chain presented by the device has been signed by the target CA cert at the control plane layer?

**Review by digimaun — COMMENTED** [review]
Review chunk.

File: sdk/iothub/azure-iot-deviceprovisioningservice/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Release History
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/CHANGELOG.md:1]: This package being rooted in sdk/deviceprovisioningservice/ is more appropriate IMO

Comment (digimaun) [reply sdk/iothub/azure-iot-deviceprovisioningservice/CHANGELOG.md:1]: sdk/deviceprovisioningservice**s** actually exists for java/js/dotnet sdk mono repos but missing in Python and I do see the conflicting takes where azure-mgmt-iothubprovisioningservices exists in sdk/iothub...

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# coding=utf-8
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# Code generated by Microsoft (R) AutoRest Code Generator.
+# Changes may cause incorrect behavior and will be lost if the code is regenerated.
+# --------------------------------------------------------------------------
+
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py:8]: Is there an apiview.dev review link available for the package?

File: sdk/iothub/azure-iot-deviceprovisioningservice/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+
+## 1.0.0b1 (2023-04-30)
+
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/CHANGELOG.md:4]: Reminder to add at least you and me as codeowners once root sdk path is confirmed.

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/auth.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/auth.py:6]: This module can be marked as private, `_auth.py`, useful public attributes like `generate_sas_token` can be exposed in deviceprovisioningservice/\_\_init__.py  \_\_all__

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    API for connecting to, and conducting operations on a Device Provisioning Service instance
+
+    :param str endpoint: The HTTP endpoint of the Device Provisioning Service instance
+    :param credential: The credential type used to authenticate with the Device Provisioning Service instance
+    """
+
+    def __init__(
+        self,
+        endpoint: str,
+        credential: Optional["TokenCredential"] = None,
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py:48]: theoretically credential on client init should be able to also support ....AzureNamedKeyCredential, AzureSasCredential and TokenCredential

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    # pylint: disable=unused-import,ungrouped-imports
+    from azure.core.credentials import TokenCredential
+
+
+class ProvisioningServiceClient(object):  # pylint: disable=client-accepts-api-version-keyword
+    """
+    API for connecting to, and conducting operations on a Device Provisioning Service instance
+
+    :param str endpoint: The HTTP endpoint of the Device Provisioning Service instance
+    :param credential: The credential type used to authenticate with the Device Provisioning Service instance
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py:42]: We should also explain why `None` for credential can work i.e. if the endpoint url contains a sas token

File: sdk/iothub/azure-iot-deviceprovisioningservice/README.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+# Initialize client
+client = ProvisioningServiceClient.from_connection_string(connection_string="<connection_string>")
+
+# Load certificate contents
+certificate = open("certificate.pem", "rt", encoding="utf-8")
+cert_contents = certificate.read()
+
+# Create x509 enrollment group
+client.enrollment_groups.create_or_update(
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioningservice/README.md:133]: Being able to setup initial device state i.e. desired tags and desired properties (for either group or individual enrollment) is also part of the hero scenarios.

Comment (digimaun) [reply sdk/iothub/azure-iot-deviceprovisioningservice/README.md:133]: We should also describe reprovisioning policy + sample of customizing would be nice.

**Review by digimaun — CHANGES_REQUESTED** [review]

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py (hunk 9 above)
Comment (digimaun) [reply sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py:42]: SAS URL is not supported, http auth header only...so credential is not optional for the service.

File: sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py (hunk 5 above)
Comment (c-ryan-k) [reply sdk/iothub/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py]: https://apiview.dev/Assemblies/Review/90240b0f26d946da9472062c5b6e72e3

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/README.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+# Create a symmetric key individual enrollment with reprovisioning policy
+client.individual_enrollments.create_or_update(
+    id="<enrollment_id>",
+    enrollment = {
+        "registrationId": "<enrollment_id>"
+        "attestation": {
+            "type": "symmetricKey",
+        },
+        "deviceId": "<device_id>",
+        "reprovision_policy": reprovision_policy
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/README.md:154]: key is in snake case, should it be reprovisionPolicy?

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        ~azure.core.credentials.AzureSasCredential or
+        ~azure.core.credentials.TokenCredential
+    """
+
+    def __init__(
+        self,
+        endpoint: str,
+        credential: Union[
+            "TokenCredential", "AzureNamedKeyCredential", "AzureSasCredential"
+        ],
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py:56]: We need to support an optional api_version keyword-only argument of type string with default value being the latest non-preview API version understood by the client.

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/samples/dps_service_sample_device_registration.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+class DeviceRegistrationSamples(object):
+    connection_string = environ["AZURE_DPS_CONNECTION_STRING"]
+    id_scope = environ["AZURE_DPS_ID_SCOPE"]
+    enrollment_group_id = "sample_symmetric_enrollment_group"
+    device_id = "test-device"
+
+    def create_enrollment_group_and_register_device(self):
+        # Instantiate a DPS Service Client using a connection string
+        from azure.iot.deviceprovisioningservice import ProvisioningServiceClient
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/samples/dps_service_sample_device_registration.py:38]: Any non-connection string samples?

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/samples/dps_service_sample_enrollment_groups.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        `openssl req -nodes -new -x509 -out enrollment-cert.pem --subj="/CN=Provisioning Service SDK Test Cert"`
+USAGE: python dps_service_sample_enrollment_groups.py
+    Set the environment variables with your own values before running the sample:
+    1) AZURE_DPS_CONNECTION_STRING - the connection string to your DPS instance.
+    2) AZURE_DPS_ENROLLMENT_CERT_PATH - Path to your certificate
+"""
+
+from os import environ
+
+
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/samples/dps_service_sample_enrollment_groups.py:25]: We need to show how error handling works with the clients.

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
+# coding=utf-8
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/client.py:1]: This client module should be private (_client.py) where the ProvisioningServiceClient is only exposed via deviceprovisioningservice \_\_init\_\_.\_\_all\_\_ attribute, similar should apply to aio version (in its corresponding \_\_init\_\_), further simplifying the public surface area.

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+from ._api_version import ApiVersion
+from ._auth import generate_sas_token
+from ._client import ProvisioningServiceClient
+from ._generated import VERSION
+from ._util import parse_connection_string
+
+__all__ = [
+    "ProvisioningServiceClient",
+    "generate_sas_token",
+    "parse_connection_string",
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py:18]: We discussed that we should start with not exposing `parse_connection_string`.

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/_generated/operations/_operations.py [hunk 17]
```diff
@@ -0,0 +1,6904 @@
...
+    @overload
+    def query(
+        self,
+        query_specification: JSON,
+        *,
+        x_ms_max_item_count: Optional[int] = None,
+        x_ms_continuation: Optional[str] = None,
+        content_type: str = "application/json",
+        **kwargs: Any
+    ) -> List[JSON]:
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/_generated/operations/_operations.py:2063]: We talked about supporting azure.core.ItemPaged[JSON] as return type, this should hide x_ms_max_items_count and x_ms_continuation

Comment (digimaun) [reply sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/_generated/operations/_operations.py:2063]: Could be opportunity to leverage the patch pattern against the generated client.

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
+# coding=utf-8
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/__init__.py:1]: We discussed package rename azure-iot-deviceprovisioningservice -> azure-iot-deviceprovisioning

File: sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/_client.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+from ._generated import ProvisioningServiceClient as GeneratedProvisioningServiceClient
+from ._generated._version import VERSION
+from ._util import parse_connection_string
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import,ungrouped-imports
+    from azure.core.credentials import TokenCredential
+
+
+class ProvisioningServiceClient(
```
Comment (digimaun) [inline sdk/deviceprovisioningservices/azure-iot-deviceprovisioningservice/azure/iot/deviceprovisioningservice/_client.py:41]: We discussed renaming `ProvisioningServiceClient` -> `DeviceProvisioningClient`

File: sdk/iothub/azure-iot-deviceprovisioning/README.md [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+)
+
+# Get device registration status for a particular device
+state = client.device_registration_state.get(
+    id="<device_id>"
+)
+```
+
+
+## Troubleshooting
```
Comment (vilit1) [inline sdk/iothub/azure-iot-deviceprovisioning/README.md:239]: very empty troubleshooting. I would expect the most common issues being authentication errors and/or I tried to add my device and I don't see the registration.

File: sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_api_version.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from enum import Enum
+
+from azure.core import CaseInsensitiveEnumMeta
+
+
+class ApiVersion(str, Enum, metaclass=CaseInsensitiveEnumMeta):
+    V2021_10_01 = "2021-10-01"
```
Comment (vilit1) [inline sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_api_version.py:12]: is this actually the newest one? Apparently for control plane, there was one public api (2022-02-05) that we skipped (cause no one told us and I think it had no new features)

Comment (digimaun) [reply sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_api_version.py:12]: It's the latest dataplane API included in the [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/deviceprovisioningservices/data-plane) repo.

For control plane there is an even [newer one](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/stable/2022-12-12) than the one you mentioned.

Comment (vilit1) [reply sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_api_version.py:12]: 😠

File: sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_client.py [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+        # Generate base client
+        self._runtime_client = GeneratedDeviceProvisioningClient(
+            credential=credential,  # type: ignore
+            endpoint=endpoint,
+            pipeline=self._pipeline,
+            api_version=api_version,
+            **kwargs,
+        )
+
+        self.individual_enrollment: IndividualEnrollmentOperations = (
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_client.py:105]: We discussed changing from individual_enrollment -> enrollment. We can do in a following PR if need be.

File: sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_generated/_client.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+        """
+
+        request_copy = deepcopy(request)
+        request_copy.url = self._client.format_url(request_copy.url)
+        return self._client.send_request(request_copy, **kwargs)
+
+    def close(self) -> None:
+        self._client.close()
+
+    def __enter__(self) -> "DeviceProvisioningClient":
```
[the remark this answers is not in the document]
Comment (c-ryan-k) [reply sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/_generated/_client.py]: This should be resolved by removing the client wrapper and directly exposing the patched `_client.py`

File: sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/aio/operations/_patch.py [hunk 24]
```diff
@@ -0,0 +1,875 @@
+# ------------------------------------
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioning/azure/iot/deviceprovisioning/aio/operations/_patch.py:1]: When I look at the commit in APIview it shows query returning `ItemPaged[JSON]` and no x_ms_continuation or x_ms_max_item_count parameters (expected) but multiple overloads of the existing non-optimized operation are exposed. Is this intended?

image

File: sdk/iothub/azure-iot-deviceprovisioning/sdk_packaging.toml [hunk 25]
```diff
@@ -0,0 +1,8 @@
+[packaging]
+package_name = "azure-iot-deviceprovisioning"
+package_pprint_name = "IoT Device Provisioning Service"
```
Comment (digimaun) [inline sdk/iothub/azure-iot-deviceprovisioning/sdk_packaging.toml:3]: Should `package_pprint_name` be "IoT Device Provisioning Client" or "IoT Device Provisioning Service Client"?

**Review by digimaun — APPROVED** [review]

**Pull request merged**